Automattic/pocket-casts-android
# Add Nova Launcher integration

## Description

This is an entry point for Nova Launcher integration. It sends subscribed podcasts to the Launcher.

When the feature flag is enabled we register two observers that sync data.

* `NovaLauncherSyncLifecycleObserver` that is lifecycle based observer registered in the process lifecycle. Whenever app goes to the background it enqueues a sync job. If there is this kind of job enqueued or active already the request is ignored.
* `NovaLauncherSyncWorker` registered as a periodic work to be executed every 6 hours.

When the feature flag is disabled all observers are unregistered.

## Testing Instructions

1. Install this version of the Nova Launcher: p1715685124487099-slack-C028JAG44VD
2. Open Nova Launcher. You don't have to set is a default launcher app but make sure you can use it.
3. Install Pocket Casts.
4. Open the app.
5. Subscribe to some podcasts.
6. Minimize the app.
7. You should see logs similar to these ones.
```
BgTask: NovaLauncherOneOffSyncWorker (Worker ID: 1cfb4398-1b4a-4f58-b483-b93dcf370409) - Enqueued Nova Launcher one-off sync
BgTask: NovaLauncherOneOffSyncWorker (Worker ID: 1cfb4398-1b4a-4f58-b483-b93dcf370409) - Staring Nova Launcher sync
BgTask: NovaLauncherOneOffSyncWorker (Worker ID: 1cfb4398-1b4a-4f58-b483-b93dcf370409) - Nova Launcher sync complete. Success: [Subscribed podcasts: 24], Failure: []
```
8. If you notice the message below in the logs. Maximize and minimize the app again. It can happen if Nova fails to fetch verification signatures.
```
Nova Launcher sync failed
java.lang.SecurityException: Failed to verify Source calling package au.com.shiftyjelly.pocketcasts.debug
```
9. Go to Nova Launcher and search for Pocket Casts or Pocket Debug.
10. You should see your subscribed podcasts enlisted.
11. Tapping on any of the podcasts should open that podcast in the app.
12. Open `App Inspection` and check `Background Task Inspector` tab.
13. You should see two `NovaLauncherSyncWorker` workers. One non-periodic and executed, and the other periodic and enqueued.
14. Go to `Beta features` in the app and disable `Integrate Pocket Cats with Nova Launcher`.
15. Minimize the app.
16. Check the `Background Task Inspector` again.
17. `NovaLauncherSyncWorker` workers should be now cancelled.
18. Enable `Integrate Pocket Cats with Nova Launcher` and kill the app.
19. Uninstall Nova Launcher.
20. Open the PC app and minimize it.
21. There should be no active `NovaLauncherSyncWorker` workers in the `Background Task Inspector`.

## Screenshots or Screencast 

![a](https://github.com/Automattic/pocket-casts-android/assets/30936061/be922ba1-1d07-49b8-96a8-ead2eff8c0ef)

## Checklist
 
#### I have tested any UI changes...
